core/src/main/java/hivemall/mf/CofactorModel.java

```diff
@@ -0,0 +1,629 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package hivemall.mf;
+
+import hivemall.fm.Feature;
+import hivemall.utils.math.MathUtils;
+import hivemall.utils.math.MatrixUtils;
+import org.apache.commons.math3.linear.ArrayRealVector;
+import org.apache.commons.math3.linear.Array2DRowRealMatrix;
+import org.apache.commons.math3.linear.RealMatrix;
+import org.apache.commons.math3.linear.RealVector;
+import org.apache.commons.math3.linear.SingularValueDecomposition;
+
+import javax.annotation.Nonnegative;
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.*;
+
+public class CofactorModel {
+
+    public enum RankInitScheme {
+        random /* default */, gaussian;
+
+        @Nonnegative
+        protected float maxInitValue;
+        @Nonnegative
+        protected double initStdDev;
+
+        @Nonnull
+        public static CofactorModel.RankInitScheme resolve(@Nullable String opt) {
+            if (opt == null) {
+                return random;
+            } else if ("gaussian".equalsIgnoreCase(opt)) {
+                return gaussian;
+            } else if ("random".equalsIgnoreCase(opt)) {
+                return random;
+            }
+            return random;
+        }
+
+        public void setMaxInitValue(float maxInitValue) {
+            this.maxInitValue = maxInitValue;
+        }
+
+        public void setInitStdDev(double initStdDev) {
+            this.initStdDev = initStdDev;
+        }
+
+    }
+
+    private static final int EXPECTED_SIZE = 136861;
+    @Nonnegative
+    protected final int factor;
+
+    // rank matrix initialization
+    protected final RankInitScheme initScheme;
+
+    @Nonnull
+    private double globalBias;
+
+    // storing trainable latent factors and weights
+    private Map<String, RealVector> theta;
+    private Map<String, RealVector> beta;
+    private Map<String, Double> betaBias;
```

Comment:
Please use `Object2DoubleMap<String> betaBias` instead to reduce memory consumption.